pkg/controller/v1alpha2/applicationconfiguration/render.go

```diff
@@ -421,8 +419,154 @@ func addDataOutputsToDAG(dag *dependency.DAG, outs []v1alpha2.DataOutput, obj *u
 	}
 }
 
-func addDataInputsToDAG(dag *dependency.DAG, ins []v1alpha2.DataInput, obj *unstructured.Unstructured, attaches []unstructured.Unstructured) {
+func (r *components) handleDependency(ctx context.Context, w *Workload, acc v1alpha2.ApplicationConfigurationComponent, dag *dag) ([]v1alpha2.UnstaifiedDependency, error) {
+	uds := make([]v1alpha2.UnstaifiedDependency, 0)
+
+	unsatsified, err := r.handleDataInput(ctx, acc.DataInputs, dag, w.Workload)
+	if err != nil {
+		return nil, errors.Wrapf(err, "handleDataInput for workload (%s/%s) failed", w.Workload.GetNamespace(), w.Workload.GetName())
+	}
+	if len(unsatsified) != 0 {
+		uds = append(uds, unsatsified...)
+		w.Unready = true
+	}
+
+	for i, ct := range acc.Traits {
+		trait := w.Traits[i]
+		unsatsified, err := r.handleDataInput(ctx, ct.DataInputs, dag, &trait.Object)
+		if err != nil {
+			return nil, errors.Wrapf(err, "handleDataInput for trait (%s/%s) failed", trait.Object.GetNamespace(), trait.Object.GetName())
+		}
+		if len(unsatsified) != 0 {
+			uds = append(uds, unsatsified...)
+			trait.Unready = true
+		}
+	}
+	return uds, nil
+}
+
+func makeUnsatisfiedDependency(obj *unstructured.Unstructured, s *dagSource, in v1alpha2.DataInput) v1alpha2.UnstaifiedDependency {
+	return v1alpha2.UnstaifiedDependency{
+		From: v1alpha2.DependencyFromObject{
+			TypedReference: runtimev1alpha1.TypedReference{
+				APIVersion: s.ObjectRef.APIVersion,
+				Kind:       s.ObjectRef.Kind,
+				Name:       s.ObjectRef.Name,
+			},
+			FieldPath: s.ObjectRef.FieldPath,
+		},
+		To: v1alpha2.DependencyToObject{
+			TypedReference: runtimev1alpha1.TypedReference{
+				APIVersion: obj.GetAPIVersion(),
+				Kind:       obj.GetKind(),
+				Name:       obj.GetName(),
+			},
+			FieldPaths: in.ToFieldPaths,
+		},
+	}
+}
+
+func (r *components) handleDataInput(ctx context.Context, ins []v1alpha2.DataInput, dag *dag, obj *unstructured.Unstructured) ([]v1alpha2.UnstaifiedDependency, error) {
+	uds := make([]v1alpha2.UnstaifiedDependency, 0)
 	for _, in := range ins {
-		dag.AddSink(in.ValueFrom.DataOutputName, obj, attaches, in.ToFieldPaths)
+		s, ok := dag.Sources[in.ValueFrom.DataOutputName]
+		if !ok {
+			return nil, errors.Wrapf(ErrDataOutputNotExist, "DataOutputName (%s)", in.ValueFrom.DataOutputName)
+		}
+		val, ready, err := r.checkSourceReady(ctx, s)
+		if err != nil {
```

Comment:
so this means that a "ready" source can be fliped to be "unready" due to network or some other errors. I think we at least should put a TODO here to fine-tune that user experience in the next PR.

Reply:
This is the reason we put the logic into reconcile and let controller runtime deal with error handling.
If such transient error happens, the AppConfig status will show it and from users' point of view they should be able to see it and wait for the network hiccup to pass away.